server/pipeline/create.go

```diff
@@ -37,109 +36,65 @@ func Create(ctx context.Context, _store store.Store, repo *model.Repo, pipeline
+	forgeYamlConfigs, configFetchErr := configFetcher.Fetch(ctx)
+
 	if configFetchErr != nil {
 		log.Debug().Str("repo", repo.FullName).Err(configFetchErr).Msgf("cannot find config '%s' in '%s' with user: '%s'", repo.Config, pipeline.Ref, repoUser.Login)
-		pipeline.Started = time.Now().Unix()
-		pipeline.Finished = pipeline.Started
-		pipeline.Status = model.StatusError
-		pipeline.Error = fmt.Sprintf("pipeline definition not found in %s", repo.FullName)
-	} else if parseErr != nil {
+		return nil, persistPipelineWithErr(ctx, _store, pipeline, repo, repoUser, fmt.Sprintf("pipeline definition not found in %s", repo.FullName))
+	}
+
+	pipelineItems, parseErr := parsePipeline(_store, pipeline, repoUser, repo, forgeYamlConfigs, nil)
+	if parseErr != nil {
 		log.Debug().Str("repo", repo.FullName).Err(parseErr).Msg("failed to parse yaml")
-		pipeline.Started = time.Now().Unix()
-		pipeline.Finished = pipeline.Started
-		pipeline.Status = model.StatusError
-		pipeline.Error = fmt.Sprintf("failed to parse pipeline: %s", parseErr.Error())
-	} else {
-		setGatedState(repo, pipeline)
+		return nil, persistPipelineWithErr(ctx, _store, pipeline, repo, repoUser, fmt.Sprintf("failed to parse pipeline: %s", parseErr.Error()))
 	}
 
+	if len(pipelineItems) == 0 {
+		log.Debug().Str("repo", repo.FullName).Msg(ErrFiltered.Error())
+		return nil, ErrFiltered
+	}
+
+	setGatedState(repo, pipeline)
+
 	err = _store.CreatePipeline(pipeline)
 	if err != nil {
-		msg := fmt.Sprintf("failure to save pipeline for %s", repo.FullName)
-		log.Error().Err(err).Msg(msg)
-		return nil, fmt.Errorf(msg)
+		msg := fmt.Errorf("failure to save pipeline for %s", repo.FullName)
+		log.Error().Err(err).Msg(msg.Error())
+		return nil, msg
 	}
 
+	pipeline = setPipelineStepsOnPipeline(pipeline, pipelineItems)
+
 	// persist the pipeline config for historical correctness, restarts, etc
+	var configs []*model.Config
 	for _, forgeYamlConfig := range forgeYamlConfigs {
-		_, err := findOrPersistPipelineConfig(_store, pipeline, forgeYamlConfig)
+		config, err := findOrPersistPipelineConfig(_store, pipeline, forgeYamlConfig)
 		if err != nil {
 			msg := fmt.Sprintf("failure to find or persist pipeline config for %s", repo.FullName)
 			log.Error().Err(err).Msg(msg)
 			return nil, fmt.Errorf(msg)
 		}
+		configs = append(configs, config)
 	}
-
-	if pipeline.Status == model.StatusError {
-		if err := publishToTopic(ctx, pipeline, repo); err != nil {
-			log.Error().Err(err).Msg("publishToTopic")
-		}
-
-		updatePipelineStatus(ctx, pipeline, repo, repoUser)
-
-		return pipeline, nil
-	}
-
-	pipeline, pipelineItems, err := createPipelineItems(ctx, _store, pipeline, repoUser, repo, forgeYamlConfigs, nil)
-	if err != nil {
-		msg := fmt.Sprintf("failure to createPipelineItems for %s", repo.FullName)
+	if err := persistPipelineConfigs(_store, configs, pipeline.ID); err != nil {
```

Comment:
Was a bit confused about `findOrPersistPipelineConfig` and `persistPipelineConfigs` first. The second is basically the linkage of an existing config (in this case prev created one) to a pipeline, right? Do you maybe know better names for them to prevent confusion?

Reply:
Maybe sth like:

```suggestion
        // link pipeline to persisted configs
	if err := persistPipelineConfigs(_store, configs, pipeline.ID); err != nil {
```

Reply:
That would make the code way more self-explanatory as well 😅